track_analyzing/log_parser.cpp

```diff
@@ -0,0 +1,168 @@
+      return expectedId;
+
+    routing::NumMwmId result = routing::kFakeNumMwmId;
+    m2::RectD const rect = MercatorBounds::RectByCenterXYAndSizeInMeters(point, 1);
+    m_mwmTree->ForEachInRect(rect, [&](routing::NumMwmId numMwmId) {
+      if (m2::RegionsContain(GetBorders(numMwmId), point))
+      {
+        result = numMwmId;
+        return;
+      }
+    });
+
+    return result;
+  }
+
+private:
+  vector<m2::RegionD> const & GetBorders(routing::NumMwmId numMwmId) const
+  {
+    auto it = m_borders.find(numMwmId);
+    CHECK(it != m_borders.cend(), ());
+    return it->second;
+  }
+
+  shared_ptr<m4::Tree<routing::NumMwmId>> m_mwmTree;
+  unordered_map<routing::NumMwmId, vector<m2::RegionD>> m_borders;
+};
+}  // namespace
+
+namespace track_analyzing
+{
+LogParser::LogParser(shared_ptr<routing::NumMwmIds> numMwmIds,
+                     unique_ptr<m4::Tree<routing::NumMwmId>> mwmTree, string const & dataDir)
+  : m_numMwmIds(move(numMwmIds)), m_mwmTree(move(mwmTree)), m_dataDir(dataDir)
+{
+  CHECK(m_numMwmIds, ());
+  CHECK(m_mwmTree, ());
+}
+
+void LogParser::Parse(string const & logFile, MwmToTracks & mwmToTracks) const
+{
+  UserToTrack userToTrack;
+  ParseUserTracks(logFile, userToTrack);
+  SplitIntoMwms(userToTrack, mwmToTracks);
+}
+
+void LogParser::ParseUserTracks(string const & logFile, UserToTrack & userToTrack) const
+{
+  my::Timer timer;
+
+  std::ifstream stream(logFile);
+  CHECK(stream.is_open(), ("Can't open file", logFile));
+
+  std::regex const base_regex(R"(.*(DataV0|CurrentData)\s+aloha_id\s*:\s*(\S+)\s+.*\|(\w+)\|)");
+  std::unordered_set<string> usersWithOldVersion;
+  uint64_t linesCount = 0;
+  size_t pointsCount = 0;
+
+  for (string line; getline(stream, line); ++linesCount)
+  {
+    std::smatch base_match;
+    if (!std::regex_match(line, base_match, base_regex))
+      continue;
+
+    CHECK_EQUAL(base_match.size(), 4, ());
+
+    string const version = base_match[1].str();
+    string const userId = base_match[2].str();
+    string const data = base_match[3].str();
+    if (version != "CurrentData")
+    {
+      CHECK_EQUAL(version, "DataV0", ());
+      usersWithOldVersion.insert(userId);
+      continue;
+    }
+
+    auto const packet = ReadDataPoints(data);
+    if (!packet.empty())
+    {
+      Track & track = userToTrack[userId];
+      track.insert(track.end(), packet.cbegin(), packet.cend());
+    }
+
+    pointsCount += packet.size();
+  };
+
+  LOG(LINFO, ("Tracks parsing finished, elapsed:", timer.ElapsedSeconds(), "seconds, lines:",
+              linesCount, ", points", pointsCount));
+  LOG(LINFO, ("Users with current version:", userToTrack.size(), ", old version:",
+              usersWithOldVersion.size()));
+}
+
+void LogParser::SplitIntoMwms(UserToTrack const & userToTrack, MwmToTracks & mwmToTracks) const
+{
+  my::Timer timer;
+
+  PointToMwmId const pointToMwmId(m_mwmTree, *m_numMwmIds, m_dataDir);
+
+  for (auto const & it : userToTrack)
```

Comment:
Optional: rename it to kv. For me "it" sounds like "iterator", while "kv" sounds like "key value".